core/drivers/cdns_uart.c

```diff
@@ -0,0 +1,99 @@
+/*
+ * Copyright (c) 2016, Xilinx Inc.
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice,
+ * this list of conditions and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ * this list of conditions and the following disclaimer in the documentation
+ * and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+#include <compiler.h>
+#include <drivers/cdns_uart.h>
+#include <io.h>
+
+#define BIT(n)		(1 << n)
+
+#define CDNS_UART_CONTROL		0
+#define CDNS_UART_MODE			4
+#define CDNS_UART_IEN			8
+#define CDNS_UART_IRQ_STATUS		0x14
+#define CDNS_UART_CHANNEL_STATUS	0x2c
+#define CDNS_UART_FIFO			0x30
+
+#define CDNS_UART_CONTROL_RXRES		BIT(0)
+#define CDNS_UART_CONTROL_TXRES		BIT(1)
+#define CDNS_UART_CONTROL_RXEN		BIT(2)
+#define CDNS_UART_CONTROL_TXEN		BIT(4)
+
+#define CDNS_UART_MODE_8BIT		(0 << 1)
+#define CDNS_UART_MODE_PARITY_NONE	(0x4 << 3)
+#define CDNS_UART_MODE_1STP		(0 << 6)
+
+#define CDNS_UART_CHANNEL_STATUS_TFUL	BIT(4)
+#define CDNS_UART_CHANNEL_STATUS_TEMPTY	BIT(3)
+#define CDNS_UART_CHANNEL_STATUS_REMPTY	BIT(1)
+
+#define CDNS_UART_IRQ_RXTRIG		BIT(0)
+#define CDNS_UART_IRQ_RXTOUT		BIT(8)
+
+void cdns_uart_flush(vaddr_t base)
+{
+	while (!(read32(base + CDNS_UART_CHANNEL_STATUS) & CDNS_UART_CHANNEL_STATUS_TEMPTY))
```

Comment:
On Mon, 2016-05-30 at 01:06:47 -0700, Jérôme Forissier wrote:
> > +#define CDNS_UART_CONTROL_TXEN		BIT(4)
> > +
> > +#define CDNS_UART_MODE_8BIT		(0 << 1)
> > +#define CDNS_UART_MODE_PARITY_NONE	(0x4 << 3)
> > +#define CDNS_UART_MODE_1STP		(0 << 6)
> > +
> > +#define CDNS_UART_CHANNEL_STATUS_TFUL	BIT(4)
> > +#define CDNS_UART_CHANNEL_STATUS_TEMPTY	BIT(3)
> > +#define CDNS_UART_CHANNEL_STATUS_REMPTY	BIT(1)
> > +
> > +#define CDNS_UART_IRQ_RXTRIG		BIT(0)
> > +#define CDNS_UART_IRQ_RXTOUT		BIT(8)
> > +
> > +void cdns_uart_flush(vaddr_t base)
> > +{
> > +	while (!(read32(base + CDNS_UART_CHANNEL_STATUS) & CDNS_UART_CHANNEL_STATUS_TEMPTY))
> 
> Line too long

I deliberately chose to violate the 80 columns as I find the code easier
to read this way. It's easy to fix, if you insist, but I'd like to keep
it this way.